kernel/src/log_segment.rs

```diff
@@ -213,17 +219,168 @@ impl LogSegment {
+    /// Each sidecar file is read and an iterator of sidecar file batches is returned
+    fn process_sidecars(
+        parquet_handler: Arc<dyn ParquetHandler>,
+        log_root: Url,
+        batch: &dyn EngineData,
+        skip_sidecar_search: bool,
+    ) -> DeltaResult<Option<impl Iterator<Item = DeltaResult<Box<dyn EngineData>>> + Send>> {
+        // If the schema does not contain add/remove actions or we have a multi-part checkpoint,
+        // we return early as sidecar files are not needed.
+        if skip_sidecar_search {
+            return Ok(None);
+        }
+
+        // Visit the rows of the checkpoint batch to extract sidecar file references
+        let mut visitor = SidecarVisitor::default();
+        visitor.visit_rows_of(batch)?;
+
+        // If there are no sidecar files, return early
+        if visitor.sidecars.is_empty() {
+            return Ok(None);
+        }
+
+        let sidecar_files: Vec<_> = visitor
+            .sidecars
+            .iter()
+            .map(|sidecar| Self::sidecar_to_filemeta(sidecar, &log_root))
+            .try_collect()?;
+
+        let sidecar_read_schema = get_log_add_schema().clone();
+
+        // Read the sidecar files and return an iterator of sidecar file batches
+        Ok(Some(parquet_handler.read_parquet_files(
+            &sidecar_files,
+            sidecar_read_schema,
+            None,
+        )?))
+    }
+
+    /// Convert a Sidecar record to a FileMeta.
+    ///
+    /// This helper first builds the URL by joining the provided log_root with
+    /// the "_sidecars/" folder and the given sidecar path.
+    ///
+    /// To catch a possible error earlier, if the sidecar.path is relative then
+    /// it is checked to ensure it is simply a file name (no directory components).
+    fn sidecar_to_filemeta(sidecar: &Sidecar, log_root: &Url) -> DeltaResult<FileMeta> {
+        // If sidecar.path is relative (does not contain "://"), require that it is
+        // just a file name. This will catch cases like "test/test/example.parquet".
+        if !sidecar.path.contains("://") {
+            let path = Path::new(&sidecar.path);
+            // If there is any parent (i.e. directory component), consider it invalid.
+            if let Some(parent) = path.parent() {
+                // If the parent is not empty, then we have extra components.
+                if parent != Path::new("") {
+                    return Err(Error::Generic(format!(
+                        "Sidecar path '{}' is invalid: sidecar files must be in the `_delta_log/_sidecars/` directory as a file name only",
+                        sidecar.path
+                    )));
+                }
+            }
+        }
+        // Now build the full location by joining log_root, "_sidecars/", and the given path.
+        let location = log_root.join("_sidecars/")?.join(&sidecar.path)?;
+        Ok(FileMeta {
+            location,
```

Comment:
single-use value considered annoying if it doesn't help with linebreaks or indentation?
```suggestion
        Ok(FileMeta {
            location: log_root.join("_sidecars/")?.join(&sidecar.path)?,
```